### PR TITLE
Adding consul capability to connect via tls

### PR DESCRIPTION
**Pull request by YKlausz**
Hello everyone,

I would like add the tls capability for consul. 
Before this update I was not able to connect to a secured consul with a given key,ca and cert files.

@brian-brazil , I think it is for you

### CODE REVIEW

File: discovery/consul/consul.go [hunk 1]
```diff
@@ -92,6 +93,26 @@ type Discovery struct {
...
+	var wrapper *http.Client
+	if &conf.TLSConfig != nil {
+		tls := &consul.TLSConfig{
+			Address:conf.Server,
+			CAFile:conf.TLSConfig.CAFile,
+			CertFile:conf.TLSConfig.CertFile,
+			KeyFile:conf.TLSConfig.KeyFile,
+			InsecureSkipVerify:conf.TLSConfig.InsecureSkipVerify,
+		}
+		consultls, err := consul.SetupTLSConfig(tls)
```
Comment (brian-brazil) [inline discovery/consul/consul.go]: This is duplicating the logic we already have, let;s just do this ourselves as we do elsewhere.

File: discovery/consul/consul.go [hunk 2]
```diff
@@ -27,6 +27,8 @@ import (
 	"golang.org/x/net/context"
 
 	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/httputil"
+	"net/http"
```
Comment (brian-brazil) [inline discovery/consul/consul.go]: You need to run make format

File: discovery/consul/consul.go [hunk 3]
```diff
@@ -92,6 +94,13 @@ type Discovery struct {
 
 // NewDiscovery returns a new Discovery for the given config.
 func NewDiscovery(conf *config.ConsulSDConfig) (*Discovery, error) {
+	tls, err := httputil.NewTLSConfig(conf.TLSConfig)
+	if err != nil {
+		return nil, err
+	}
+	transport := &http.Transport{TLSClientConfig: tls}
+	wrapper := &http.Client{Transport: transport}
```
Comment (juliusv) [inline discovery/consul/consul.go]: Just `wrapper` -> `client` would be a more common naming.

**Comment by juliusv** [comment]
@YKlausz Ping - seems like besides my nit, the only outstanding TODO here is a `make format` run?

File: discovery/consul/consul.go (hunk 2 above)
Comment (grobie) [inline discovery/consul/consul.go]: Gofmt (which is calledby our make target) doesn't move package imports between import blocks. Calling make/gofmt won't help here.

@YKlausz please move that import up to the other imports of stdlib and keep the list ordered (which gofmt will enforce).

File: discovery/consul/consul.go (hunk 3 above)
Comment (YKlausz) [inline discovery/consul/consul.go]: Variable name client is already used in the fonction

**Comment by YKlausz** [comment]
Thank you and now Is it good ?

File: discovery/consul/consul.go [hunk 4]
```diff
@@ -24,9 +25,9 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/log"
 	"github.com/prometheus/common/model"
-	"golang.org/x/net/context"
```
Comment (grobie) [inline discovery/consul/consul.go]: So personally I don't care, but the current three blocks had a reason: 1) imports from stdlib 2) imports from external packages 3) imports from internal (prometheus/prometheus/...) packages. You merged the second and third group together.

**Comment by Gouthamve** [comment]
I think this fixes #2365.

**Comment by juliusv** [comment]
Merging, we can fix the import groups later if we care.

**Comment by juliusv** [comment]
@Gouthamve Thanks for making the connection, I will mention in that issue and close it!

**Comment by juliusv** [comment]
Filed an issue for documenting the new options: https://github.com/prometheus/docs/issues/681 - anyone, feel free to grab it.

**Pull request merged**